llvm/include/llvm/Support/ThreadPool.h

```diff
@@ -227,7 +265,7 @@ class ThreadPool {
 class ThreadPoolTaskGroup {
```

Comment:
Only minor nit, if the user wants to customize their `ThreadPoolTaskGroup` (to add something like a condition variable for their use case let's say) they can't -- we're using the concrete implementation in `mlir/include/mlir/IR/Threading.h`. Do you see any avenues for this?

Reply:
I'm not sure I understand the use-case, but this seems like a different feature anyway?

Reply:
Wouldn't the implemention for `ThreadPoolTaskGroup` come in hand with the one for `ThreadPool`? I feel if someone implements the `ThreadPoolInterface` they would want something similar for the groups? The use case is like I said: the new implementation would want to store something in the `ThreadPoolTaskGroup`.

Reply:
Get the point of @aganea; the `ThreadPoolTaskGroup` is a bit tricky to reason when it comes to customized extensibility as the concrete implementation is injected to the interface itself. IMHO I think that this construct can sit on top of the `ThreadPoolInterface` itself (as it kind of does now) without semantically injecting itself in the pool. Someone can implement this through some API that returns a trackable construct; e.g. a `TaskID` or some kind of `Future` interface from the `asyncEnqueue`. Not sure if that's worth the effort though.

Reply:
> Wouldn't the implemention for ThreadPoolTaskGroup come in hand with the one for ThreadPool? 

Right now the only thing the TaskGroup provide is a unique ID in the form of its own address.
I don't quite get what I would change in the implementation? Would you want the group to customize its own `wait()`? Right now it dispatches to the threadpool implementation: `wait(ThreadPoolTaskGroup &Group)`.

Customizing the task group would require to change the pattern at every use site as well: https://github.com/llvm/llvm-project/blob/0e0bee26e7f33c065eebef9a674b2f19bb156414/mlir/include/mlir/IR/Threading.h#L69-L70

Instead of:
```
  llvm::ThreadPool &threadPool = context->getThreadPool();
  llvm::ThreadPoolTaskGroup tasksGroup(threadPool);
```
we would need to use the pool as a factory somehow:
```
  llvm::ThreadPool &threadPool = context->getThreadPool();
  std::unique_ptr<llvm::ThreadPoolTaskGroup> tasksGroup = threadPool.createTaskGroup();
```

I'm not opposed to this, but that still seems like a separate change, and one that would better be done when there is somehow with the need for this.

Reply:
That point was raised by the OP in the Discourse thread (allow for custom implementations of `ThreadPoolTaskGroup`) and I think it deserved a clear answer, and I do agree now to your point. Until there's no clear use-case, the behavior of `ThreadPoolTaskGroup` will remain as is, deferred to the `ThreadPool` implementation.

Reply:
Btw, I still think that passing a `ThreadPoolTaskGroup` for the sole purpose of a uid is a bit brittle as you bind the interface with this construct but internally to the `ThreadPool` no member function of it is called?!
We could change for example this with a `std::string` with probably minimal changes to any call site.
Apologies if that is not right but I am not familiar with the code.

Reply:
> We could change for example this with a std::string or uint64_t with probably minimal changes to any call site.

How would you get a unique id at runtime if not for the address of the object?

Reply:
Sure you can convert the address of the object to a string or int that is fine; it is not the best but not worse than what is there. You can do this inside the `ThreadPoolTaskGroup` which is essentially a RAII wrapper.
What you get is probably a cleaner API IMHO and don't impose to the implementations to have to consume a struct that by definition loops around the interface.

Reply:
> What you get is probably a cleaner API IMHO
 
 I don't quite get why it would cleaner actually? This makes is for a weaker interface that does not guarantee any uniqueness and you lose type safety as well.
 I must missing your point here.

Reply:
So the way I see this is:

I have a public API that looks like:
```
  virtual void asyncEnqueue(std::function<void()> Task, ThreadPoolTaskGroup *Group) = 0;
 // or similar
```

My thoughts:
* `ThreadPoolTaskGroup` implementation is actually internal to the `ThreadPool`
* `ThreadPoolTaskGroup` being passed to the API but only used internally as just a UID. No member functions being called internally. Moreover, the object definition itself is a RAII wrapper on top of the ThreadPool API implementation so I find that this is an unnecessary circular dependency and coupling.
* Uniqueness is `guaranteed` through memory addressing. Someone can `break` uniqueness when going through the public `ThreadPool` API.
* Type safety; sure you don't have implicit conversions going on.

What `guarantees` to you uniqueness is essentially when you use RAII wrapper itself and the fact that calls at termination: ` ~ThreadPoolTaskGroup() { wait(); }` which will go through; process the tasks and erase the group from the `ActiveGroups`.

So what I think is that passing instead of passing `*this` you can convert the address to a string/int and pass this instead.

Eitherway, please ignore my comment if it doesn't make sense; as said not familiar with the codebase. I just find strange a bit this aspect of the API.

Reply:
> Uniqueness is "guaranteed" through memory addressing. Someone can break uniqueness when going through the public ThreadPool API.


How so?

> Type safety; sure you don't have implicit conversions going on.


There is much more than: the API is constrained to say: "it operates on an instance of a TaskGroup object".
This is a very important API: it is explicit in terms of the scope on which it operates, and the fact that it care about the **actual instance** of the TaskGroup object.


> So what I think is that instead of passing *this you can convert the address to a string/int and pass this instead; hence you don't lose any guarantees. I understand your arguments but I find this a bit more decoupled.

As I said above, I would have strong concerns with this: it is a weak API (it is open to any integer and does not carry the semantics I described above) and "unsafe", just like using `void*`. 
There is not way for a user to know what integer to pass here for example. As a user such API is highly confusing and really I'm not sure what the pattern would be other than always casting the address of the task group to an int.

Reply:
Thanks for explaining @joker-eph. Looks good to me.